### PR TITLE
[FLINK-19449][table-planner] LEAD/LAG cannot work correctly in streaming mode

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

Fix "LEAD/LAG cannot work correctly in streaming mode".

`LeadLagAggFunction` is only works in batch mode with `OffsetOverFrame`.
In streaming mode, it will output wrong results.

## Brief change log

- Add `LinkedListSerializer`.
- Fix wrong document for lead and lag.
- Pass isBounded to `AggFunctionFactory`.
- Implement `LagAggFunction`.
- Add Lag IT Case for streaming mode.

## Verifying this change

- OverAggregateITCase.testLagFunction
- OverAggregateITCase.testLeadFunction

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no) no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no) no
  - The serializers: (yes / no / don't know) no
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know) no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / no / don't know) no
  - The S3 file system connector: (yes / no / don't know) no

## Documentation

  - Does this pull request introduce a new feature? (yes / no) no

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/AggFunctionFactory.scala [hunk 1]
```diff
@@ -52,7 +52,8 @@ import scala.collection.JavaConversions._
 class AggFunctionFactory(
```
Comment (cshuo) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/AggFunctionFactory.scala:52]: minor: 
param in doc is miss matching.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/functions/aggfunctions/LagAggFunction.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            acc.buffer.removeFirst();
+        }
+    }
+
+    public void accumulate(LagAcc<T> acc, T value, int offset) throws Exception {
+        acc.offset = offset;
+        accumulate(acc, value);
+    }
+
+    public void accumulate(LagAcc<T> acc, T value, int offset, T defaultValue) throws Exception {
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/functions/aggfunctions/LagAggFunction.java:107]: This method will invoke above method.

**Comment by cshuo** [comment]
LGTM, +1

**Comment by JingsongLi** [comment]
I will split commits and merge.

**Pull request merged**